src/librustc/middle/resolve_lifetime.rs

```diff
@@ -1943,22 +1939,22 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                 trait_ref: &hir::PolyTraitRef,
                 modifier: hir::TraitBoundModifier,
             ) {
-                self.binder_depth += 1;
+                self.outer_index.shift_in(1);
                 intravisit::walk_poly_trait_ref(self, trait_ref, modifier);
-                self.binder_depth -= 1;
+                self.outer_index.shift_out(1);
             }
 
             fn visit_lifetime(&mut self, lifetime_ref: &hir::Lifetime) {
                 if let Some(&lifetime) = self.map.defs.get(&lifetime_ref.id) {
                     match lifetime {
                         Region::LateBound(debruijn, _, _) | Region::LateBoundAnon(debruijn, _)
-                            if debruijn.depth < self.binder_depth =>
+                            if debruijn < self.outer_index =>
                         {
                             self.have_bound_regions = true;
                         }
                         _ => {
                             self.lifetimes
-                                .insert(lifetime.from_depth(self.binder_depth));
```

Comment:
OK, I suspect this is where the bug is coming from. The `binder_depth` in the "before code" is equal to `index`, but (if you look at the definition of `from_depth`) you will see that used to be computing adjusting by `depth - 1`, basically.

I think the way I would want to compute the "depth" by giving two debruijn indices. In particular, we might do something like `self.outer_index.depth_from(INNERMOST)`, where we have:

```rust
impl DebruijnIndex {
    /// Assuming that these two debruijn indices are both in scope,
    /// returns the number of binding levels *between* them.
    /// This is often used if `outer` represents some starting binding
    /// and `self` represents the current innermost binding; it would then
    /// compute the number of bindings passed over so far.
    ///
    /// As the name suggests, `outer` is assumed  to be the
    /// outermost depth of the two.
    fn depth_from(self, outer: DebruijnIndex) -> usize {
        assert!(self.index >= other.index); // or return the absolute value
        self.index - other.index
    }
}
```